[Attention] Unify mamba and attention backend selection

ayushsatyam146 opened the pull request:
## Purpose

Fixes #23073. This PR unifies mamba and attention backend selection logic by removing the separate `get_mamba_attn_backend()` function and implementing the standard `Layer.get_attn_backend()` interface for all layer types.

**Fixes the issue where mamba models used duplicate backend selection logic instead of the unified approach.**

### Changes Made:
- **Deleted** `vllm/v1/attention/backends/mamba_selectors.py` to eliminate duplicate code
- **Updated** `MambaMixer` and `MambaMixer2` classes to implement `get_attn_backend()` method
- **Modified** `gpu_model_runner.py` to use the unified `get_attn_backends_for_layers()` function for all layer types
- **Removed** mamba-specific backend selection logic from model runner
- **Updated** affected model files to use the new unified interface

This makes mamba models more pluggable and follows the same pattern as regular attention layers:
- `MambaMixer.get_attn_backend()` → `Mamba1AttentionBackend`  
- `MambaMixer2.get_attn_backend()` → `Mamba2AttentionBackend`

LucasWilkinson commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2803,19 +2803,15 @@ def create_attn_groups(
-                    get_mamba_attn_backend(kv_cache_spec.mamba_type):
-                    kv_cache_group_spec.layer_names
-                }
+                layer_type = MambaBase
```
can we make a common ancestor for `MambaBase` and `Attention` maybe `AttentionLayerBase` with the abstract method `.get_attn_backend()`

killershrimp commented on tests/v1/attention/test_mamba_selectors.py:
```diff
@@ -4,22 +4,104 @@
 
-    assert backend_class is expected_backend
+@pytest.fixture
+def distributed_init():
```
maybe can use this existing fixture instead? https://github.com/vllm-project/vllm/blob/0167efe20d3d2280c3da6aea94a6f59afec5099c/tests/conftest.py#L171

ayushsatyam146 replied:
Thanks, this makes sense. I am sorry I am new to this project and didn't pay attention on this, thanks for pointing out.

heheda12345 commented on tests/v1/attention/test_mamba_selectors.py:
```diff
@@ -4,22 +4,104 @@
 
 import pytest
 
+# yapf: disable
```
can you enable it after line 9?

ayushsatyam146 replied:
I resolved it without needing to disable it. So I will remove it completely in next commit. Is that fine?

heheda12345 replied:
That's great!

Josephasafg commented:
I think this is a good and cleaner approach overall. Lets let @tdoublep / @heheda12345 give their opinion as well

ayushsatyam146 commented on tests/v1/attention/test_mamba_selectors.py:
```diff
@@ -4,22 +4,85 @@
 
 import pytest
 
+from vllm.model_executor.layers.mamba.mamba_mixer import MambaMixer
```
Can you please suggest a new good name? 😅

ayushsatyam146 replied:
renamed it to `test_attention_backends_selection.py`

ayushsatyam146 commented:
Hi @heheda12345 @LucasWilkinson @Josephasafg, this PR needed rebase because `ShortConvAttentionBackend` was added and my PR needed to accommodate it's changes as well. I have rebased it and addressed all concerns, including some new suggestions by @Josephasafg. Please review, Thanks!

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2770,7 +2773,7 @@ def get_attn_backends_for_layers(
             # layer.
             for layer_name in layer_names:
-                attn_backend = attn_layers[layer_name].get_attn_backend()
+                attn_backend = layers[layer_name].get_attn_backend()
```
nit: can you move `attn_layers = get_layers_from_vllm_config(self.vllm_config, Attention)` (line 2760 now) to the line that this variable is first used.

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2799,20 +2802,11 @@ def create_attn_groups(
-                    kv_cache_group_spec.layer_names
-                }
-            else:
+            if not isinstance(kv_cache_spec, (AttentionSpec, MambaSpec)):
```
Feel that we don't need this check now. We can assume all layers have implemented `get_attn_backend`.

ayushsatyam146 replied:
Thanks for clarifying this, Done!

heheda12345 reviewed (commented):
Thanks! Only a few small comments.

ayushsatyam146 committed:
refactor: Unify mamba and attention backend selection
    Use Layer.get_attn_backend() interface for all layer types instead of
    separate mamba-specific backend selection logic.

Signed-off-by: Ayush Satyam <ayushsatyam146@gmail.com>

heheda12345 reviewed (approved):
LGTM! Thanks for the clean-up.

LucasWilkinson reviewed (approved):
LGTM; thanks for doing this!
